### PR TITLE
ARTEMIS-1542 - AMQP message cluster-bridging fix

### PR DESCRIPTION
**Pull request by toddbaert**
Fixing and issue where AMQP messages would lose their address when being sent accross a cluster-bridge.

### CODE REVIEW

File: artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPMessage.java [hunk 1]
```diff
@@ -608,7 +608,11 @@ public AMQPMessage setAddress(String address) {
 
    @Override
    public AMQPMessage setAddress(SimpleString address) {
-      return setAddress(address.toString());
+      if (address != null) {
+        return setAddress(address.toString());
+      } else {
+        return null;
```
Comment (clebertsuconic) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPMessage.java]: shouldn't this be:


return setAddress(null);

????

**Comment by toddbaert** [comment]
Yes, looking again, I think you're right. I think the primary issue was the null argument to the beforeForward() call though. I've updated the PR.

**Comment by toddbaert** [comment]
Added fix for ambiguous method call to setAddress(null), and checkstyles.

**Comment by michaelandrepearce** [comment]
@toddbaert is there a test case for this to be added, to both prove fix, and avoid future regression?

Also could you squash up your commits into one, once test is added. Thanks.

**Comment by toddbaert** [comment]
@michaelandrepearce I'll add a test case and consolidate the commits.

**Comment by clebertsuconic** [comment]
@michaelandrepearce you need a producer and have a meessage with destination == null to make this happen.

the AMQP JMS client will set the destination on the producer.. what won't happen through the C++ client.


You would need to run the test with a C++ client.


The best would probably be to have an AMQP Example with clustered messages.

**Comment by clebertsuconic** [comment]
hmmm.. you've made some mess of your branch here.. I can't merge it...


I have an older version.. where I added an example based on yours...
I will merge that one... leave it with me

**Comment by clebertsuconic** [comment]
@toddbaert leave this with me... merging it now

**Comment by clebertsuconic** [comment]
@toddbaert thanks a lot.. merged it manually. let me know what you think.

**Comment by toddbaert** [comment]
@clebertsuconic @michaelandrepearce 

The problem seems to me to be pretty simple:

in ClusterConnectionBridge.beforeForward(Message message, SimpleString forwardingAddress), we call super.beforeForward(message, **null**). The is why we lose the address. This code path is executed when broker1 forwards an AMQP message to broker2. This is why we will only see this in a cluster, when a producer sends a message to broker1 and a consumer is listening on broker2.

**Comment by clebertsuconic** [comment]
@toddbaert that was a problem only if message.address was null.. I didn't think a cpp producer would leave it null...

it should be fixed now.

**Comment by toddbaert** [comment]
@clebertsuconic  I don't think you've address the issue. The address is set to null BY broker1 when it forwards the message to broker2. Please see ClusterConnectionBridge.beforeForward(Message message, SimpleString forwardingAddress), it call super.beforeForward(message, null), regardless of the forwarding message in the original call.

**Comment by toddbaert** [comment]
@clebertsuconic @michaelandrepearce 

Guys - please see my messages above. The primary issue is that the forwarding address is set to null, as I described.. I was in the middle of an interactive rebase when you attempted to merge (which I've now completed). New PR is here: https://github.com/apache/activemq-artemis/pull/1702.

**Comment by clebertsuconic** [comment]
@toddbaert  Ohhhh... @#$@$!@@! (bad word goes here)... I screwed up...  I had a debug thingy that I committed as part of this....

**Comment by toddbaert** [comment]
@clebertsuconic you also did not commit my proposed change to ClusterConnectionBridge.java, which is the heart of our issue.

https://github.com/apache/activemq-artemis/pull/1702

**Comment by clebertsuconic** [comment]
@toddbaert it was an accident.. didn't mean it.

**Comment by toddbaert** [comment]
I'm not sure I understand what you're referring to. Anyway, I hope that you can merge https://github.com/apache/activemq-artemis/pull/1702, as it is the main problem. If you have any concerns, let me know and I'll help however I can. You can produce the issue with my configuration and sample c# client on my original JIRA: https://issues.apache.org/jira/browse/ARTEMIS-1549

:smiley:

**Comment by clebertsuconic** [comment]
@toddbaert during my merge, I added an example... on that example.. I reverted your fix by accident... I didn't mean to setNull... sorry about that.

I will merge your PR as that should fix it.

**Comment by toddbaert** [comment]
@clebertsuconic  ahhhh OK, I understand now. Thanks!

**Pull request closed without merging**